### PR TITLE
WebRTC

### PR DESCRIPTION
**Pull request by kevinshen56714**
Build webcam feature

### CODE REVIEW

File: src/client/web/WebRTC.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import Peer from 'peerjs'
+import Network from '../services/Network'
+
+export default class WebRTC {
+  private myPeer: Peer
+  private peers = {}
+  private videoGrid = document.getElementById('video-grid')
```
Comment (DaxChen) [inline src/client/web/WebRTC.ts:7]: you can use `document.querySelector('.video-grid')` after changing these from id to class

File: src/client/web/WebRTC.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  private buttonGrid = document.getElementById('button-grid')
+  private myVideo = document.createElement('video')
+  private myStream?: MediaStream
+
+  constructor(userId: string, network: Network) {
+    this.myPeer = new Peer(userId)
+
+    // mute your own video stream (you don't want to hear to yourself)
```
Comment (DaxChen) [inline src/client/web/WebRTC.ts:15]: nit:     // mute your own video stream (you don't want to hear ~to~ yourself)

File: src/client/web/WebRTC.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import Peer from 'peerjs'
+import Network from '../services/Network'
+
+export default class WebRTC {
+  private myPeer: Peer
+  private peers = {}
```
Comment (DaxChen) [inline src/client/web/WebRTC.ts:6]: Use Map<> instead of object.
This is a tricky topic in JS, both object {} and Map<> are considered "hash map", like dictionary in python, so theoretically this is the same runtime complexity O(1) lookup. However when the number of keys gets larger, JS object performance becomes really bad compared to Map.
TL;DR: just use Map when you want a hash map.

File: src/client/web/WebRTC.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    video.addEventListener('loadedmetadata', () => {
+      video.play()
+    })
+    if (this.videoGrid) {
+      this.videoGrid.append(video)
+    }
+  }
+
+  // method to remove video stream a peer disconnects
+  deleteVideoStream(userId: string) {
```
Comment (DaxChen) [inline src/client/web/WebRTC.ts:75]: also remember to remove the reference from the hashmap here

```
// If using Map
this.peers.delete(userId)

// (FYI if using object)
delete this.peers[userId]
```

File: src/client/web/WebRTC.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    if (this.peers[userId]) this.peers[userId].close()
+  }
+
+  // method to set up mute/unmute and video on/off buttons
+  setUpButtons() {
+    const audioButton = document.createElement('button')
+    audioButton.innerHTML = 'Mute'
```
Comment (DaxChen) [inline src/client/web/WebRTC.ts:82]: `innerHTML` is dangerous and can lead to XSS, use `innerText` instead since you're not using any HTML here.

Also there's many things I can pick here such as these event listeners are never cleaned up, but since we'll be transitioning these to react it doesn't matter...

File: src/client/index.html [hunk 6]
```diff
@@ -13,9 +13,40 @@
 		*:after {
 			box-sizing: inherit;
 		}
+
```
Comment (DaxChen) [inline src/client/index.html:16]: Setup your vscode to use prettier to format HTML files on-save too

hurts my eyes to see these huge tabs instead of spaces LOL

**Review by DaxChen — APPROVED** [review]
Well done! You're getting the hang of TS!
I've approved this PR, but remember to resolve the comments, rebase the parent PR, change the target to master, before merging.

File: src/client/index.html [hunk 7]
```diff
@@ -13,9 +13,40 @@
...
+			top: 35px;
+			right: 10px;
+		}
+
+		video {
+			width: 100%;
+			height: 100%;
+			object-fit: cover;
+			border-radius: 5px;
+			border-style: groove;
```
[the remark this answers is not in the document]
Comment (kevinshen56714) [reply src/client/index.html:32]: Cool, done!

File: src/client/index.html [hunk 8]
```diff
@@ -13,9 +13,40 @@
 		*:after {
 			box-sizing: inherit;
 		}
+
+		#video-grid {
```
[the remark this answers is not in the document]
Comment (kevinshen56714) [reply src/client/index.html:17]: I see. Done.

File: src/client/index.html [hunk 9]
```diff
@@ -13,9 +13,40 @@
...
+			display: grid;
+			grid-template-columns: repeat(auto-fill, 160px);
+			grid-gap: 5px;
+			grid-auto-rows: 160px;
+			position: absolute;
+			top: 35px;
+			right: 10px;
+		}
+
+		video {
```
[the remark this answers is not in the document]
Comment (kevinshen56714) [reply src/client/index.html:27]: Cool, this for sure will be polished when React comes in. Done.

File: src/client/index.html [hunk 10]
```diff
@@ -13,9 +13,40 @@
...
+			width: 100%;
+			height: 100%;
+			object-fit: cover;
+			border-radius: 5px;
+			border-style: groove;
+			border-width: 1px;
+			border-color: rgb(229, 251, 255);
+		}
+		
+		#button-grid {
```
[the remark this answers is not in the document]
Comment (kevinshen56714) [reply src/client/index.html:37]: Done.

File: src/client/services/Network.ts [hunk 11]
```diff
@@ -48,7 +43,13 @@ export default class Network {
     // an instance removed from the players MapSchema
     this.room.state.players.onRemove = (player: IPlayer, key: string) => {
       this.events.emit('player-left', key)
+      if (this.webRTC) this.webRTC.deleteVideoStream(key)
```
[the remark this answers is not in the document]
Comment (kevinshen56714) [reply src/client/services/Network.ts:46]: I see. Done!

File: src/client/web/WebRTC.ts (hunk 1 above)
Comment (kevinshen56714) [reply src/client/web/WebRTC.ts:7]: Gotcha. Done!

File: src/client/web/WebRTC.ts (hunk 2 above)
Comment (kevinshen56714) [reply src/client/web/WebRTC.ts:15]: LOL. Bad bad Chinglish.

File: src/client/web/WebRTC.ts (hunk 3 above)
Comment (kevinshen56714) [reply src/client/web/WebRTC.ts:6]: Cool, I wanted ask you about this but forgot. Have changed it to Map.

File: src/client/web/WebRTC.ts (hunk 4 above)
Comment (kevinshen56714) [reply src/client/web/WebRTC.ts:75]: Done!

File: src/client/web/WebRTC.ts (hunk 5 above)
Comment (kevinshen56714) [reply src/client/web/WebRTC.ts:82]: Hmm... I see. Changed it to ```innerText``` and look forward to React framework :D

File: src/client/index.html (hunk 6 above)
Comment (kevinshen56714) [reply src/client/index.html:16]: OK. Got Prettier to work on HTML files now! 💯
